### ISSUE TITLE
Version parser is confused by release suffixes (rc2, beta1, ...)

### ISSUE THREAD

**Issue by vrubiolo** [lxc/lxd#8586]
# Required information

 * Distribution: Fedora
 * Distribution version: 33
 * The output of "lxc info" or if that fails
```
config:
  core.https_address: '[10.31.0.114]:8443'
  images.auto_update_interval: "0"
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 10.31.0.114:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIICHTCCAaKgAwIBAgIRAI9GcJoj2rm7XnZaImxK3bMwCgYIKoZIzj0EAwMwPDEc
    MBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEcMBoGA1UEAwwTcm9vdEB2aW5j
    ZW50LWxhcHRvcDAeFw0yMDA2MTUxMzEyMjhaFw0zMDA2MTMxMzEyMjhaMDwxHDAa
    BgNVBAoTE2xpbnV4Y29udGFpbmVycy5vcmcxHDAaBgNVBAMME3Jvb3RAdmluY2Vu
    dC1sYXB0b3AwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAAQqZwUpsb43gR+XdnKZsrWp
    bqXXmFJB1wIObCgsBq4OOkKJxnUvBNPkJcFVXQdHsfRvoFFHRJrl0Sf11HVE63fK
    sOfZbsjCzqCu2yG+XF0uMUipjla1cOPewd94Pk92EOWjaDBmMA4GA1UdDwEB/wQE
    AwIFoDATBgNVHSUEDDAKBggrBgEFBQcDATAMBgNVHRMBAf8EAjAAMDEGA1UdEQQq
    MCiCDnZpbmNlbnQtbGFwdG9whwR/AAABhxAAAAAAAAAAAAAAAAAAAAABMAoGCCqG
    SM49BAMDA2kAMGYCMQCHTnRKzLJpSDP79xkfrckaHpt/JoBYyJ7GPk1ai+FpZFiK
    L8WAgY4hq7GiAdK7DDQCMQDJFwjZChtjSnXxNySCwB/3p473LjFGItkOTzqJzh01
    uAs4htwgSK2q7IXiF85fCZ8=
    -----END CERTIFICATE-----
  certificate_fingerprint: 84d97a927ec32f4edf306d7dd6dc29f593cbb7e43fea276aa0c8c656d0a7cd38
  driver: lxc
  driver_version: 4.0.6
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.10.19-200.fc33.x86_64
  lxc_features:
    cgroup2: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    seccomp_notify: "true"
  project: default
  server: lxd
  server_clustered: false
  server_name: vrubiolo-laptop
  server_pid: 3631680
  server_version: "3.20"
  storage: dir
  storage_version: "1"
```

# Issue description

My containers on my Fedora33 host have started to fail getting IP addresses whereas it used to work before. Upon investigation, I have found the following in the lxd log:
```
Mar 19 09:22:47 vrubiolo-laptop lxd[3429151]: t=2021-03-19T09:22:47+0100 lvl=eror msg="Failed to bring up network" err="Invalid version format: \"2.84rc2\"" name=lxdbr0
```

That rc-based version string matches the dnsmasq installed on my system:
```
$ dnsmasq --version                                                                                         
Dnsmasq version 2.84rc2  Copyright (c) 2000-2021 Simon Kelley                                                                                                                                                                                 
Compile time options: IPv6 GNU-getopt DBus no-UBus no-i18n IDN2 DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth cryptohash DNSSEC loop-detect inotify dumpfile                                                                                
                                                                                                                       
This software comes with ABSOLUTELY NO WARRANTY.                                                                       
Dnsmasq is free software, and you are welcome to redistribute it                                                       
under the terms of the GNU General Public License, version 2 or 3.
```

It looks like lxd fails at parsing that version number in https://github.com/lxc/lxd/blob/master/shared/version/version.go#L19. 

Incidentally, it looks like this comes from the manufacturing side of `dnsmasq` for which the `git describe` based commands picked up two tags, one for `2.84` and one for `2.84rc2`, as seen in the `VERSION` file of the [2.84 tarball](https://thekelleys.org.uk/dnsmasq/dnsmasq-2.84.tar.xz):
```
$ pwd
/home/vrubiolo/rpmbuild/BUILD/dnsmasq-2.84
$ cat VERSION 
 (HEAD -> master, tag: v2.84rc2, tag: v2.84, origin/master, origin/HEAD)
[05:47 PM] vrubiolo@vrubiolo-laptop ~/rpmbuild/BUILD/dnsmasq-2.84
```

My workaround is to downgrade the package to 2.82 via `dnf downgrade` and my containers are working fine again:
```
$ dnsmasq --version
Dnsmasq version 2.82  Copyright (c) 2000-2020 Simon Kelley
[...]
```
I am filing this bug report as per https://discuss.linuxcontainers.org/t/lxd-dnsmasq-version-handling-code-chokes-at-2-84rc2-string-preventing-containers-dhcp-from-working/10562/2?u=vrubiolo

# Steps to reproduce

1. Install Fedora 33 and update to the latest version. You should get dnsmasq v2.84rc2
2. Try to start a container using the `lxdbr0` bridge in DHCP mode
3. Note how the container fails to pick an IP address from DHCP. The abovementioned lxd log shows that network cannot be brought up.

Additional information:
LXD is installed from https://copr.fedorainfracloud.org/coprs/ganto/lxc4/

**Comment by monstermunchkin** [comment]
>   server_version: "3.20"

You're running an old LXD version. The issue was fixed in LXD 4.7: https://github.com/lxc/lxd/commit/1fddf116b6e4222b69c7d834e59c26d7b88d7763

**Comment by vrubiolo** [comment]
Thanks @monstermunchkin for the quick update, I will see to update to 4.x then.

**Comment by vrubiolo** [comment]
I confirm the issue is fixed after upgrading to 4.1. 

For the record, here is my system info:
```
config:
  core.https_address: '[10.31.0.114]:8443'
  images.auto_update_interval: "0"
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
- network_dns_search
- container_nic_routed_limits
- instance_nic_bridged_vlan
- network_state_bond_bridge
- usedby_consistency
- custom_block_volumes
- clustering_failure_domains
- resources_gpu_mdev
- console_vga_type
- projects_limits_disk
- network_type_macvlan
- network_type_sriov
- container_syscall_intercept_bpf_devices
- network_type_ovn
- projects_networks
- projects_networks_restricted_uplinks
- custom_volume_backup
- backup_override_name
- storage_rsync_compression
- network_type_physical
- network_ovn_external_subnets
- network_ovn_nat
- network_ovn_external_routes_remove
- tpm_device_type
- storage_zfs_clone_copy_rebase
- gpu_mdev
- resources_pci_iommu
- resources_network_usb
- resources_disk_address
- network_physical_ovn_ingress_mode
- network_ovn_dhcp
- network_physical_routes_anycast
- projects_limits_instances
- network_state_vlan
- instance_nic_bridged_port_isolation
- instance_bulk_state_change
- network_gvrp
- instance_pool_move
- gpu_sriov
- pci_device_type
- storage_volume_state
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 10.31.0.114:8443
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIICHTCCAaKgAwIBAgIRAI9GcJoj2rm7XnZaImxK3bMwCgYIKoZIzj0EAwMwPDEc
    MBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEcMBoGA1UEAwwTcm9vdEB2aW5j
    ZW50LWxhcHRvcDAeFw0yMDA2MTUxMzEyMjhaFw0zMDA2MTMxMzEyMjhaMDwxHDAa
    BgNVBAoTE2xpbnV4Y29udGFpbmVycy5vcmcxHDAaBgNVBAMME3Jvb3RAdmluY2Vu
    dC1sYXB0b3AwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAAQqZwUpsb43gR+XdnKZsrWp
    bqXXmFJB1wIObCgsBq4OOkKJxnUvBNPkJcFVXQdHsfRvoFFHRJrl0Sf11HVE63fK
    sOfZbsjCzqCu2yG+XF0uMUipjla1cOPewd94Pk92EOWjaDBmMA4GA1UdDwEB/wQE
    AwIFoDATBgNVHSUEDDAKBggrBgEFBQcDATAMBgNVHRMBAf8EAjAAMDEGA1UdEQQq
    MCiCDnZpbmNlbnQtbGFwdG9whwR/AAABhxAAAAAAAAAAAAAAAAAAAAABMAoGCCqG
    SM49BAMDA2kAMGYCMQCHTnRKzLJpSDP79xkfrckaHpt/JoBYyJ7GPk1ai+FpZFiK
    L8WAgY4hq7GiAdK7DDQCMQDJFwjZChtjSnXxNySCwB/3p473LjFGItkOTzqJzh01
    uAs4htwgSK2q7IXiF85fCZ8=
    -----END CERTIFICATE-----
  certificate_fingerprint: 84d97a927ec32f4edf306d7dd6dc29f593cbb7e43fea276aa0c8c656d0a7cd38
  driver: lxc | qemu
  driver_version: 4.0.6 | 5.1.0
  firewall: nftables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "true"
    seccomp_listener: "true"
    seccomp_listener_continue: "true"
    shiftfs: "false"
    uevent_injection: "true"
    unpriv_fscaps: "true"
  kernel_version: 5.10.23-200.fc33.x86_64
  lxc_features:
    cgroup2: "true"
    devpts_fd: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "true"
    seccomp_allow_deny_syntax: "true"
    seccomp_notify: "true"
    seccomp_proxy_send_notify_fd: "true"
  os_name: Fedora
  os_version: "33"
  project: default
  server: lxd
  server_clustered: false
  server_name: vrubiolo-laptop
  server_pid: 44665
  server_version: "4.11"
  storage: dir
  storage_version: "1"
```
